### PR TITLE
Add Deinit Required rule

### PR DESCRIPTION
**Pull request by BenStaveleyTaylor**
Classes are required to have a deinit method.

As proposed in https://github.com/realm/SwiftLint/issues/2620

This is a style to help debugging memory issues, when it is common to want to set a breakpoint at the point of deallocation. Often classes don't have a deinit, so the developer ends up having to quit, add a deinit and rebuild to proceed. If all classes have a deinit, debugging is much smoother.

### CODE REVIEW

File: Source/SwiftLintFramework/Rules/Style/DeinitRequiredRule.swift [hunk 1]
```diff
@@ -0,0 +1,76 @@
+import SourceKittenFramework
+
+public struct DeinitRequiredRule: OptInRule, ConfigurationProviderRule, AutomaticTestableRule {
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/Style/DeinitRequiredRule.swift]: what do you think about renaming it to `RequiredDeinitRule`? It'd be more in line with the existing `RequiredEnumCaseRule`.

File: Source/SwiftLintFramework/Rules/Style/DeinitRequiredRule.swift [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            }
+            """
+        ]
+    )
+
+    public init() {}
+
+    public func validate(file: File) -> [StyleViolation] {
+        let classCollector = NamespaceCollector(dictionary: file.structure.dictionary)
+        let classes = classCollector.findAllElements(of: [.class])
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/Style/DeinitRequiredRule.swift]: nitpick: you can implement `ASTRule` instead of using `NamespaceCollector`

**Comment by marcelofabri** [comment]
> This cuts out the classes NamespaceCollector but I still use one for the method instances unless there's a better way of listing all methods in a class?

It's great now, thanks!

> I confess I was unsure whether this was an Style, Idiomatic or Lint rule. I hope Style is OK.

Good point. IMO `style` is more for just code formatting. I'd use `lint` in this case. What do you think?

**Comment by BenStaveleyTaylor** [comment]
> Good point. IMO `style` is more for just code formatting. I'd use `lint` in this case. What do you think?

Sure, I've made the change. I couldn't find any guidelines for when to use the different kinds -- is this documented somewhere? It might be nice to cover it in Contributing.md? To be brutally honest, I don't think the kind adds much value to a rule definition and I'd rather it went away :-) ...

**Comment by marcelofabri** [comment]
> I couldn't find any guidelines for when to use the different kinds -- is this documented somewhere? 

Unfortunately, no. I'll try to write something.

>  To be brutally honest, I don't think the kind adds much value to a rule definition and I'd rather it went away

The hope is that it'd be more useful in the future. It could be used to group documentation (as Rules.md is already quite long) or even enable/disable a group of rules on a configuration.

Right now it serves to group rules in Xcode and it's also shown in Rules.md.

**Pull request merged**